Implemented Warehouse Pypi API Call

Allstreamer opened the pull request:
- Added log 0.4 Crate
- Added serde 1.0 Crate
- Added serde_json 1.0 Crate
- Created pypi.rs
- Created request_package_info function
- Created PypiData struct
- Added request_package_info example to main.rs

Implementaition of what was mentioned in
- #13
- #17

Allstreamer committed:
Implemented Warehouse Pypi API Call

- Added log 0.4 Crate
- Added serde 1.0 Crate
- Added serde_json 1.0 Crate
- Created pypi.rs
- Created request_package_info function
- Created PypiData struct
- Added request_package_info example to main.rs

John15321 commented:
Please change target to the develop branch

Allstreamer committed:
Merge branch 'develop' into pypi-api

Allstreamer commented:
Solved the merge conflicts and retargeted to develop

Allstreamer committed:
Formatting & Linting

- Solved clippy suggestions
- Ran cargo fmt

Allstreamer commented:
Fixed Formating & Linting

John15321 commented:
When writing the name pypi with capital letters make sure its "PyPI" not Pypi or PyPi that also includes structs

John15321 commented:
Please add docstrings to all structs, methodsz functions etc

John15321 commented:
Please add tests. Both usage unit tests in the example section of the docstrings and negative cases in a test module and some integration tests to check if the cli works.
 I know it may be complicated but that's why we don't have a lot of written code rn. I have to research mock modules etc we want to use so try ur best for now

John15321 commented:
Additionally i don't see any issues connected to this PR that explain what's being done here. So please create one

Allstreamer committed:
Updated pypi module

- Added Docstrings
- Renamed PypiData to PyPIData
- Added unit tests

Allstreamer commented:
Implementaion of #17

Allstreamer commented:
#17

Allstreamer commented:
Should now be mergable

Allstreamer committed:
Rename check_pytorch_download_v1

- Renamed check_pytorch_download_v1 to check_numpy_download_name_v1

Allstreamer commented:
Had a name mixed up where i was getting info for numpy but the test was named pytorch

Allstreamer committed:
Updated pypi.rs to include negative test case

- Added Negative test case

(Haven't added others for .get() calls since those will be replaced soon with more specific structs)

Allstreamer committed:
Update pypi.rs

- Fixed some doc strings

Allstreamer committed:
Added PyPIPackageInfo & PyPIPackageDownloadInfo

- Created PyPIPackageInfo & PyPIPackageDownloadInfo Structs
- Updated unit-tests, examples & doc examples

Kiwifuit commented:
You could have your `PyPIPackageInfo` look like this since we wont need most of the info that the JSON API gives:
```rust
struct PyPIPackageInfo {
    version: String, // Or a tuple of MAJOR, MINOR, PATCH assuming that every python package uses semver
    digest: String,
    url: String,
    python_version: String, // Again, semver tuple can be used here
    size: u8,               // Can be helpful in progress bars
    filename: String,
}
```

Allstreamer commented:
Not every package uses semver (see package "AWRS" which has the version 2019.8.14.1152)

Allstreamer commented:
I'll probably rewrite it that way but first i need to find out everything that rust-pip could need

Allstreamer commented:
Yeah, don't worry too much! I'm still a long ways away from being happy with the code in this pull request!

Still need a way to calculate python version requirements from strings such as:
```>=2.7,!=3.0.*,!=3.1.*,!=3.2.*,!=3.3.*```

A way of managing package versions and release information,
Since not every release that's listed in the api has information.

and so on!

So i'll be crafting away at this pull request for a while!

Allstreamer commented:
```rust
#[derive(Debug, Serialize, Deserialize, PartialEq)]
pub struct PyPIPackageRelease {
    version: String,
    filename: String,
    md5_digest: String,
    required_python: String,
    size: u32,
    url: String,
}

#[derive(Debug, Serialize, Deserialize, PartialEq)]
pub struct PyPIPackageData {
    name: String,
    stable_version: String,
    releases: Vec<PyPIPackageRelease>
}
```

Something like this but with the versions replaced with some kind of struct is what i'm thinking of @Kiwifuit

John15321 commented:
I remember finding crates that solve server for you maybe you can find something. Also please mark this PR as draft if that's the case

Allstreamer commented:
> I remember finding crates that solve server for you maybe you can find something. Also please mark this PR as draft if that's the case

Sadly PyPi allows arbitrary strings and not just semver

Allstreamer commented:
I've been looking into how pip manages versions and it looks like they have two systems:
- Their own version system in this format: `Something`
- And a legacy system which allows abitrary strings:
```python
class LegacyVersion(_BaseVersion):
    def __init__(self, version: str) -> None:
        self._version = str(version)
        self._key = _legacy_cmpkey(self._version)

        warnings.warn(
            "Creating a LegacyVersion has been deprecated and will be "
            "removed in the next major release",
            DeprecationWarning,
        )
```
Since as it says above that the legacy system will be removed with the next release i think i should just implement their current version system and throw an `Err` when reciving an abitrary string

John15321 commented:
Yeah in general i don't plan on supporting anything legacy. This is supposed to be bloatless pip for the future

whispersofthedawn commented on src/pypi.rs:
```diff
@@ -0,0 +1,126 @@
+
+/// Download stats
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct PyPIPackageDownloadInfo {
```
```suggestion
pub struct PypiPackageDownloadInfo {
```
The typical Rust convention is to keep acronyms with the first letter in uppercase, and the rest in lowercase. This would also need to be updated in the rest of the file.

Allstreamer replied:
John wanted this formatting and I think matching the Offical naming looks better.

whispersofthedawn commented on src/pypi.rs:
````diff
@@ -0,0 +1,126 @@
+///
+/// # Example
+/// ```
+/// use rust-pip::PyPI::request_package_info;
````
I'm pretty sure this would error with invalid syntax. Have you tried running the doc tests? That should be `cargo test --doc`.

dawnofmidnight commented:
> > I remember finding crates that solve server for you maybe you can find something. Also please mark this PR as draft if that's the case
> 
> Sadly PyPi allows arbitrary strings and not just semver 

I think this would be better in a separate PR, personally.

Allstreamer committed:
Fixed Documentation
